### PR TITLE
Add zero'ing

### PR DESCRIPTION
**Pull request by thorstenhater**
# Add specialised zero method to rangeutils.

Zeroing current arrays can take substantial amounts of time in simple models. Thus
add zero leveraging memset which is slightly faster and communicates intentions
a bit better.

### CODE REVIEW

File: arbor/util/rangeutil.hpp [hunk 1]
```diff
@@ -9,6 +9,7 @@
 #include <iterator>
 #include <ostream>
 #include <numeric>
+#include <cstring>
```
Comment (brenthuisman) [inline arbor/util/rangeutil.hpp:12]: Are you sure? I don't see anything using this.

**Review by brenthuisman — APPROVED** [review]

File: arbor/util/rangeutil.hpp (hunk 1 above)
Comment (thorstenhater) [reply arbor/util/rangeutil.hpp:12]: std::memset?

Comment (brenthuisman) [reply arbor/util/rangeutil.hpp:12]: :eyes:

File: arbor/util/rangeutil.hpp [hunk 2]
```diff
@@ -82,6 +83,16 @@ void fill(Seq&& seq, const V& value) {
     std::fill(canon.begin(), canon.end(), value);
 }
 
+// Zero a container, specialised for contiguous sequences
+// i.e.: Array, Vector, String.
+
+template <typename Seq, typename = std::enable_if_t<sequence_traits<Seq&&>::is_contiguous>>
```
[the remark this answers is not in the document]
Comment (thorstenhater) [reply arbor/util/rangeutil.hpp:89]: Oh how I wish for concepts 😢

File: arbor/util/rangeutil.hpp [hunk 3]
```diff
@@ -82,6 +83,18 @@ void fill(Seq&& seq, const V& value) {
     std::fill(canon.begin(), canon.end(), value);
 }
 
+// Zero a container, specialised for contiguous sequences
+// i.e.: Array, Vector, String.
+
+template <typename Seq,
+          typename = std::enable_if_t<sequence_traits<Seq&&>::is_contiguous>,
+          typename = std::is_trivially_copyable<typename sequence_traits<Seq&&>::value_type>>
```
Comment (AdhocMan) [inline arbor/util/rangeutil.hpp:91]:
```suggestion
          typename = std::enable_if_t<
              sequence_traits<Seq &&>::is_contiguous &&
              std::is_trivially_copyable_v<typename sequence_traits<Seq &&>>>>
```
The `value_type` of `is_trivially_copyable` is just a type that always exists, so using it this way has no useful effect. You'll want to use `value` or the `_v` definition.
And yes, concepts would be nice!

File: arbor/util/rangeutil.hpp (hunk 1 above)
Comment (AdhocMan) [inline arbor/util/rangeutil.hpp:12]:
```suggestion
#include <cstring>
#include <type_traits>
```

**Review by AdhocMan — CHANGES_REQUESTED** [review]

**Review by AdhocMan — APPROVED** [review]
lgtm

**Pull request merged**